### DISCUSSION TITLE
How to download student submissions for assignment?

### DISCUSSION THREAD

**Discussion by drscotthawley** [ucfopen/canvasapi#634] [Q&A] [answered]
Hi!  I've enjoyed playing around with `canvasapi` and am able to get some basic info from my university's canvas instance.  

I apologize for broaching what I imagine is a very common question, but I haven't seen it covered on the Docs page https://canvasapi.readthedocs.io/en/stable/examples.html#assignments
...or elsewhere (see "PS-" at the bottom of this message).

My question is: How do I (programmatically, in Python) download the files that students have submitted for an assignment? 

If I execute code such as 

```python
canvas = Canvas(API_URL, API_KEY)
course = canvas.get_course(course_id)
assn_id = 1234  
assignment = course.get_assignment(assn_id)
subs = assignment.get_submissions()
```
.... it gives me a `PaginatedList` for `subs`, which I print it out...

```python
for s in subs:
    print(s)
```
I see a set of strings that read like `<assignment_id>-<student_id>`. 

...But I'm not sure what to do after that!   Since the list `subs` isn't  subscriptable, I'm not sure how to investigate what it contains. 

Can anyone help?  Thanks. 

PS- In looking for answers, so far I've only found: 

* nothing in these Discussions
*  [these GUI-based instructions](https://community.canvaslms.com/t5/Instructor-Guide/How-do-I-download-all-student-submissions-for-an-assignment/ta-p/760) which is not what I want. 
* [this Canvas LMS Community answer about an API target for listing all assignments](https://community.canvaslms.com/t5/Canvas-Question-Forum/Is-there-an-API-call-that-get-a-list-of-all-submissions-for-one/m-p/209670), but I don't understand how to make use of the answer given. 
* The docs for [Submission](https://canvasapi.readthedocs.io/en/stable/submission-ref.html) don't seem to suggest anything for downloading files that students submitted. 
* The docs for [Assignment](https://canvasapi.readthedocs.io/en/stable/assignment-ref.html) have a `upload_to_submission` target but don't seem to have any corresponding `download_from_submssion`. :-/

**Comment by drscotthawley** [comment]
...Ok. Insead I print `__dir__()` as in:
```python
for s in subs:
    print(s.__dir__())
```
Then I see a list of attributes that reads
```
['_requester', 'id', 'body', 'url', 'grade', 'score', 'submitted_at', 'assignment_id', 'user_id', 'user_id_date', 'submission_type', 'workflow_state', 'grade_matches_current_submission', 'graded_at', 'grader_id', 'attempt', 'cached_due_date', 'cached_due_date_date', 'excused', 'late_policy_status', 'points_deducted', 'grading_period_id', 'extra_attempts', 'posted_at', 'redo_request', 'custom_grade_status_id', 'sticker', 'late', 'missing', 'seconds_late', 'entered_grade', 'entered_score', 'preview_url', 'course_id', 'attachments', '__module__', '__init__', '__str__', 'create_submission_peer_review', 'delete_submission_peer_review', 'edit', 'get_submission_peer_reviews', 'mark_read', 'mark_unread', 'upload_comment', '__doc__', '__getattribute__', '__repr__', 'set_attributes', '__dict__', '__weakref__', '__new__', '__hash__', '__setattr__', '__delattr__', '__lt__', '__le__', '__eq__', '__ne__', '__gt__', '__ge__', '__reduce_ex__', '__reduce__', '__getstate__', '__subclasshook__', '__init_subclass__', '__format__', '__sizeof__', '__dir__', '__class__']
```
....not sure where the attached file is?  If I print out `attachments` it's an empty list, i.e. `[]`. :-/ 

--- 
Alternatively, by doing a test submission with "Test Student", I'm able to see that the syntax of the URL that forms for the file download is 
`https://<university>.instructure.com/courses/<course_id>/assignments/<assignment_id>/submissions/<student_id>?download=1150248&inline=1`
Not sure what that "1150248" is for, or how to reliably get that info. 
...or how to use that URL with my API_KEY to actually get the file from within Python or Terminal.

**Reply by whampt** [reply]
Thank you so very much for writing out how to get a list of available attributes for a Canvas object!!
I assume that this is not included in the CanvasAPI documentation as Paginated List is considered Python knowledge not specific to this project.

**Comment by drscotthawley** [comment, marked as answer]
Oooo! I got it!   I could close this thread now, but perhaps I should leave it open in case wiser people want to chime in?

```python
ls = 0
for s in subs:
    print(f"-----\nls={ls}")
    if len(s.attachments) > 0:
        thefile = s.attachments[-1]
        thefile.download(f"a{s.assignment_id}_u{s.user_id}_{thefile}")
    ls+= 1
```
This will download my students' submitted  files to the current directory. Inspecting them locally shows that the files are intact! :-) 


Admins, feel free to close if you like.
